PetscDMWrapper: Fieldsplit Capabilities (Replaces #1568)

pbauman opened the pull request:
I guess if you force push to a branch you can't reopen the pull request. This replaces #1568. 

I believe I've addressed all the comments from the previous version, in particular, this should support all the finite element types, including SCALARs. I've tested this on both `ReplicatedMesh` and `DistributedMesh`, both in serial and parallel, including cases with more processors than elements. I've tried Lagrange and Hierarchic elements and can reproduce the same convergence behavior as using the "old style" libMesh field split interface for Stokes and Navier-Stokes problems. I think I got it right, but would appreciate a look to see if I did anything obviously stupid. The only assumption that I'm not asserting here (I think) and for which I don't know how to assert within this code is that Node::n_dofs() and Elem::n_dofs() will not both count the same dof for the same variable. Originally I was thinking that elements were only nodal or elemental dof-oriented, but looks like Hierarchic elements associates dofs at the nodes for low order and then the interior for higher order.

The only catch, as described in 814081f, is that this currently will only do the correct thing for fieldsplit when using node-major ordering. @knepley is updating the PETSc side to support var-major ordering automatically at which point we can use this for both var-major and node-major orderings. We'll simply need to uncomment the currently commented petsc function call.

@bboutkov should rebase on this branch; that branch will add geometric multigrid. At that point, we'll want to open the discussion on integration with `PetscNonlinearSolver` (beyond the current `PetscDiffSolver`).

roystgnr commented:
Copying our telecon topic into the ticket so we don't forget it:

> both var-major and node-major orderings

There's also variable-group-major ordering, which is our *default*.  For the short-term the user can turn off automatic VariableGroup construction and the PetscDM stuff can scream-and-die if it hasn't been turned off, but in the medium term we ought to support it; the VariableGroup optimization saves a noticeable amount of memory and probably some CPU too.

roystgnr commented on src/base/dof_map.C:
```diff
@@ -2322,8 +2322,7 @@ void DofMap::SCALAR_dof_indices (std::vector<dof_id_type> & di,
   // If it's not in the local indices
-  if (dof_index < this->first_dof() ||
-      dof_index >= this->end_dof())
+  if (!this->local_index(dof_index))
```
There's a dozen other places we could make this change too, while we're at it.

pbauman replied:
I did at least admit my laziness in the commit log. :) I'll do a quick grep later to see if I can find a few more.

roystgnr commented on include/solvers/petsc_dm_wrapper.h:
```diff
@@ -43,6 +49,44 @@ class PetscDMWrapper
+private:
+
+  //! Vector of DMs for all grid levels
+  std::vector<std::unique_ptr<DM>> _dms;
```
They put "DM" into the global namespace??  :frowning:

pbauman replied:
Sadly yes. And `TS` if/when we add their time solvers.

roystgnr commented on src/solvers/petsc_dm_wrapper.C:
```diff
@@ -91,6 +91,11 @@ void PetscDMWrapper::build_section( const System & system, PetscSection & sectio
 
+  // Sanity checking at least that total n_dofs match
+#ifndef NDEBUG
+  this->check_section_n_dofs_match(system, section);
```
Should this just return total_n_dofs and then you can save 2 lines by wrapping it in libmesh_assert_equal_to()?

pbauman replied:
Good point. Comment is actually wrong, it should be `n_local_dofs`, but I can make the suggested change, that will be better, thanks.

pbauman commented:
> There's also variable-group-major ordering, which is our default. For the short-term the user can turn off automatic VariableGroup construction and the PetscDM stuff can scream-and-die if it hasn't been turned off, but in the medium term we ought to support it; 

Correct, I'm glad you brought this up because I obviously forgot. I'll push a change to 814081f with commented code for that check just so we don't forget.

> the VariableGroup optimization saves a noticeable amount of memory and probably some CPU too.

I'll have to work with @knepley on interacting with their PetscSection to set the correct offsets in that case.

I would suggest as follows (after addressing other @roystgnr comments): 

1. we merge this since it's a new feature that works with node-major ordering for arbitrary elements so it's a nice new feature and we can start hooking in GMG.

2. We merge in change when PETSc function is correct and still throw error when multiple variable group handling is detected. (We'll need to test we do the correct thing still in that case). I expect this to be relatively short term.

3. Implement handling for variable group case after 1 and 2. This will be medium term because I (or someone) will need to sit down with @knepley to work with their API or they need to document for us how to do what we want.

roystgnr commented on src/solvers/petsc_dm_wrapper.C:
```diff
@@ -99,6 +99,71 @@ void PetscDMWrapper::build_section( const System & system, PetscSection & sectio
+
+  // This is the vector of dof indices coming from other processors
+  // TODO: We do a stupid copy here since we can't convert an
+  //       unsigned int* to PetscInt*
```
Don't we static_assert that dof_id_type and PetscInt are the same size?  You ought to be able to reinterpret_cast send_list.data(), right?  Maybe add another static_assert here if you're feeling paranoid.

We do a copy in PetscVector::localize, but that's just because we're combining the send list with the local indices anyway.

pbauman replied:
Damn. I really need to get more familiar with C++11.

roystgnr commented:
Code looks good, but not having specific test coverage of it scares me.  I'm not sure how to best *add* a "scream and die when Roy breaks your code" quality test when PETSc is enabled, but we need one.

pbauman commented:
> Code looks good, 

Thanks! The documentation was clear enough? (You didn't complain so I'll assume it was unless you reply otherwise.) I tried to document the hell out of this because 3 years from now I'd look at the code again and go "WTF?" without it.

> but not having specific test coverage of it scares me. 

Agreed. I'd planned on adding GRINS tests for this for sure after this was merged (the ones I used to test during the development of this feature), but honestly, I'm not sure exactly *how* to test this. The only way that I knew it was broken in the first place was the linear solver took one extra iteration compared to the "old style fieldsplit" when the PetscSection didn't have the correct ordering... I might be able to make it more sensitive with out such a sledge hammer of a solver and then look into the LinearSolutionMonitor and check to make sure the number of iterations didn't change, but then even that will depend on the number of processors.

I might be able to extract the ordering in the PetscSection from the PetscSectionView (or something) and make sure the ordering is correct there.

I'm certainly open to other suggestions in this regard.

> I'm not sure how to best add a "scream and die when Roy breaks your code" quality test when PETSc is enabled, but we need one.

Not just Roy, but any PETSc changes too!

pbauman commented:
OK, I've done the following:

1. Squashed in update to the n_local_dof sanity checking into 585d025 so that it is a one line check now.
2. Squashed in reinterpret_cast<>(const_cast<>()) (:-O) into e0e31f6 so we don't copy the send_list before handing it to PETSc (which will copy it again).
3. Squashed more commented out code into c730ea7 for future updates in PETSc function so that var-major ordering will be supported but will error out on var-group-major ordering. I'm sure will need to update it again at that time for cases like there's a var group, but only 1, so we can just use node-major ordering, etc., but this was more just so we don't forget.
4. Added b5cdbe8 which got most (if not all) of the places where we could use `DofMap::local_index` instead of code like `if(i < first_dof() || i >= last_dof()).` I did not squash this in on purpose. @roystgnr or @jwpeterson or anyone please look at that commit to make sure I have the logic right on all the checks. I checked several times, but it would be real easy to miss a `!` in there.

This works for my test cases for Stokes (Q2/Q1) and Incompressible Navier Stokes (Q2/Q1) + stabilization (Q1/Q1, Q2/Q2, equal order hierachic up to order 4) on replicated and distributed mesh 1--16 processors (including a Stokes case where there were more processors than elements).

So I believe this is ready to go once I add some kind of testing, modulo other comments folks want to make. AGain, I'm thinking just dropping in a uncoupled Laplace test and checking convergence on a really small mesh and making sure we converge in 1 iteration. If the test seems too cumbersome, I'll at least put up a GRINS PR with some regression checks on the test cases I used to develop the feature and then we can merge (will do that anyway, but would make priority).

@bboutkov, please rebase on this and let's get a PR ready to go for the GMG.

knepley commented:
On Wed, Jul 18, 2018 at 9:13 AM Paul T. Bauman <notifications@github.com>
wrote:

> OK, I've done the following:
>
>    1. Squashed in update to the n_local_dof sanity checking into 585d025
>    <https://github.com/libMesh/libmesh/commit/585d025bd09b54ecc1127201833ec229caef55d2>
>    so that it is a one line check now.
>    2. Squashed in reinterpret_cast<>(const_cast<>()) (:-O) into e0e31f6
>    <https://github.com/libMesh/libmesh/commit/e0e31f6eaf471d96ce2ea5ce0f11e0b007b07c7a>
>    so we don't copy the send_list before handing it to PETSc (which will copy
>    it again).
>    3. Squashed more commented out code into c730ea7
>    <https://github.com/libMesh/libmesh/commit/c730ea763fd2e5b6ecce8ece7f25d86e556e6ef5>
>    for future updates in PETSc function so that var-major ordering will be
>    supported but will error out on var-group-major ordering. I'm sure will
>    need to update it again at that time for cases like there's a var group,
>    but only 1, so we can just use node-major ordering, etc., but this was more
>    just so we don't forget.
>    4. Added b5cdbe8
>    <https://github.com/libMesh/libmesh/commit/b5cdbe816bde6c4f4152e06b228bc48178405996>
>    which got most (if not all) of the places where we could use
>    DofMap::local_index instead of code like if(i < first_dof() || i >=
>    last_dof()). I did not squash this in on purpose. @roystgnr
>    <https://github.com/roystgnr> or @jwpeterson
>    <https://github.com/jwpeterson> or anyone please look at that commit
>    to make sure I have the logic right on all the checks. I checked several
>    times, but it would be real easy to miss a ! in there.
>
> This works for my test cases for Stokes (Q2/Q1) and Incompressible Navier
> Stokes (Q2/Q1) + stabilization (Q1/Q1, Q2/Q2, equal order hierachic up to
> order 4) on replicated and distributed mesh 1--16 processors (including a
> Stokes case where there were more processors than elements).
>
> So I believe this is ready to go once I add some kind of testing, modulo
> other comments folks want to make. AGain, I'm thinking just dropping in a
> uncoupled Laplace test and checking convergence on a really small mesh and
> making sure we converge in 1 iteration. If the test seems too cumbersome,
> I'll at least put up a GRINS PR with some regression checks on the test
> cases I used to develop the feature and then we can merge (will do that
> anyway, but would make priority).
>
> @bboutkov <https://github.com/bboutkov>, please rebase on this and let's
> get a PR ready to go for the GMG.
>
Another possible test is to use full Schur complement with exact
factorizations and the full Schur complement preconditioner with LU. Then
everything should go in one iterate.

  Matt

> —
>
You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/libMesh/libmesh/pull/1799#issuecomment-405925427>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAjoiaLCHV1u_bqvo4GF1uzXTjvEP2meks5uHzRegaJpZM4VP9Fh>
> .
>


-- 
What most experimenters take for granted before they begin their
experiments is infinitely more interesting than any results to which their
experiments lead.
-- Norbert Wiener

https://www.cse.buffalo.edu/~knepley/ <http://www.caam.rice.edu/~mk51/>

roystgnr commented:
:+1: once a test's in, unless anyone else objects

roystgnr commented:
ETA on tests?  I'm working on some stuff that I'd like to merge after this to avoid conflicts, so if test coverage won't be done soon then I'd just as soon get the code in now and get the tests in a week later.

pbauman commented:
> ETA on tests? 

Sorry I didn't get this done before being without wifi access and cell signal. Back in civilization now.

> I'm working on some stuff that I'd like to merge after this to avoid conflicts, so if test coverage won't be done soon then I'd just as soon get the code in now and get the tests in a week later.

Let's go ahead and merge this then. I'd like to finish up the new test today, but my day is slowly getting destroyed, so it might be a few days before I can finish it.

balborian commented:
Were tests or examples for this feature ever implemented?
